File: src/tools/endpoints.ts
// endpoints — deduplicated method+host+path inventory across the snapshot.

import { applyFilter, buildFilter } from "../filters.js";
import { renderEndpoints } from "../format/render.js";
import type { SnapshotStore } from "../snapshot.js";

export interface EndpointsArgs {
  host?: string;
  path?: string;
  method?: string | readonly string[];
}

export async function endpoints(
  store: SnapshotStore,
  args: EndpointsArgs = {},
): Promise<string> {
  const snap = await store.get();
  const f = buildFilter({ host: args.host, path: args.path, method: args.method });
  const entries = applyFilter(snap.entries, f);

  const counts = new Map<string, number>();
  // Maintain a parallel map for the tuple to avoid splitting strings on render.
  const tuples = new Map<string, [string, string, string]>();

  for (const e of entries) {
    const pathNoQuery = (e.request.path || "").split("?", 1)[0]!;
    const key = `${e.request.method}|${e.request.host}|${pathNoQuery}`;
    counts.set(key, (counts.get(key) ?? 0) + 1);
    if (!tuples.has(key)) tuples.set(key, [e.request.method, e.request.host, pathNoQuery]);
  }

  const rows: [string, string, string, number][] = [];
  for (const [key, count] of counts) {
    const t = tuples.get(key)!;
    rows.push([t[0], t[1], t[2], count]);
  }
  rows.sort((a, b) => {
    if (b[3] !== a[3]) return b[3] - a[3];
    if (a[1] !== b[1]) return a[1] < b[1] ? -1 : 1;
    return a[2] < b[2] ? -1 : a[2] > b[2] ? 1 : 0;
  });
  return renderEndpoints(rows);
}

File: src/burp_mcp_lite/tools/endpoints.py
"""endpoints — deduplicated method+host+path inventory across the snapshot."""

from __future__ import annotations

from collections import Counter
from typing import Optional, Sequence, Union

from ..filters import build_filter, apply_filter
from ..format.render import render_endpoints
from ..snapshot import SnapshotStore


async def endpoints(
    store: SnapshotStore,
    *,
    host: Optional[str] = None,
    path: Optional[str] = None,
    method: Union[None, str, Sequence[str]] = None,
) -> str:
    snap = await store.get()
    fil = build_filter(host=host, path=path, method=method)
    entries = apply_filter(snap.entries, fil)

    counts: Counter[tuple[str, str, str]] = Counter()
    for e in entries:
        # Strip the query string for dedup; query params are noise for inventory.
        path_no_query = e.request.path.split("?", 1)[0]
        counts[(e.request.method, e.request.host, path_no_query)] += 1

    rows = sorted(
        [(m, h, p, n) for (m, h, p), n in counts.items()],
        key=lambda r: (-r[3], r[1], r[2]),
    )
    return render_endpoints(rows)
